### ISSUE TITLE
buildGatewayListeners: no gateways for router

### ISSUE THREAD

**Issue by adamzhoul** [istio/istio#8595]
**Describe the bug**
    gateway get 0 listeners. which leads to all my traffic goes down. this is awful.

**Expected behavior**
   get 2 listeners.(80 443)

**Steps to reproduce the bug**

   1、  run ingressgateway as daemonSet.
   2、  run other svc as daemonset too.  Make sure they scheduled to the same host.
   3、  create gateway crd.
   4、  check the log of pilot .

        info	ads	LDS: PUSH for node:istio-ingressgateway-cgbj5.istio-system addr:"10.233.65.0:54112" listeners:0

part of my gatway crd:
   ```
    spec:
    selector:
      istio: ingressgateway
    servers:
    - hosts:
      - '*'
      port:
        name: http
        number: 80
        protocol: HTTP
    - hosts:
      - '*'
      port:
        name: http2
        number: 443
        protocol: HTTP2

   ```

   This bug doesn't happen 100%.

**Version**
istio 1.0.1

**Is Istio Auth enabled or not?**
no

**Environment**
centos

**Cluster state**
healthy

**Code**
 
  To solve the problem, I look into the code and find something .

![image](https://user-images.githubusercontent.com/7430439/45299792-3285d980-b53f-11e8-84c1-80c1adfb6019.png)
   
   I labled my ingress "istio=ingressgateway". so my crd can be bind correctly.
   However, when i follow the pilot log. And I find that "workloadLabels" doesn't have that .

![image](https://user-images.githubusercontent.com/7430439/45299712-02d6d180-b53f-11e8-9a92-c2ace9c1e20b.png)

   so, i check the func : "env.GetProxyServiceInstances(node)" . and find below:

![image](https://user-images.githubusercontent.com/7430439/45300808-ed16db80-b541-11e8-9ccd-471dff8f05a4.png)

   now ,  time to check :  "c.pods.labelsByIP(ea.IP)"
![image](https://user-images.githubusercontent.com/7430439/45301178-fbb1c280-b542-11e8-85ed-3e0cd9dc5403.png)

![image](https://user-images.githubusercontent.com/7430439/45301225-184dfa80-b543-11e8-81ed-59144bdd01b3.png)

From pictures above, i think  "pc.getPodByIP(addr)" get the wrong pod. and it can not get the correct pod. since the data struct.

**Comment by adamzhoul** [comment]
forget to mention.
network mode of daemonset is "host". that is why they share the same IP.

**Comment by adamzhoul** [comment]
I don't know if anyone else has encountered this problem.
to avoid this. I have to remove the "selector" part of gateway config.

after doing that,  can make sure get the resource even has incorrect labels. because gateway has no label selector part.

I have to admit that this is not a regular or right way, but it solved my problem.
wish someone can help fix this problem the right way.

**Comment by jiuchen1986** [comment]
@adamzhoul

**Comment by jiuchen1986** [comment]
@adamzhoul

**Comment by jiuchen1986** [comment]
@adamzhoul 
Same issue. Guess istio not recommends to use the host network.

**Comment by adamzhoul** [comment]
@jiuchen1986 
Do you have any other solution? since mine doesn't look grace enough.

**Comment by andraxylia** [comment]
cc @costinm @duderino

**Comment by andraxylia** [comment]
@adamzhoul would you be so kind to take the latest release of Istio 1.0.3 ? The 1.0.1 had some issues.

What is you scale in terms of number of services and pods?

**Comment by costinm** [comment]
I don't think 1.0.2 can handle this either (1.0.3 is still in progress). It's a larger change to support host network - but we do support NodePort.

It is not impossible - just not something we have right now. We had several discussions on moving away
from using IP to identify workloads, there are many good reasons to do so - but we need to prioritize,
and realistically such large/risky changes can take few releases to finalize ( including migration, upgrade, etc). So I wouldn't expect it in 1.1 - but maybe after, depending on resources and other work. 

We may be able to use a workaround, specifically for Gateways - but I'm not sure yet how hard it is
and if anyone can volunteer to work on this.

**Comment by adamzhoul** [comment]
if the host network is not supported, how you guys deploy gateway, may I ask? 
it needs to receive traffic from outside, and traffic only hit physical machine.

if we use NodePort, we can't add new port dynamically, right?

**Comment by adamzhoul** [comment]
> We had several discussions on moving away
from using IP to identify workloads, 

@costinm 
is this still in the discussion? do we have a link, it can help we learn more info about this. 
thanks, ：）

**Comment by adamzhoul** [comment]
@jiuchen1986  
 in your case,  we should place a proxy(like nginx). 
we have many similar cases too. like multi https certificates.

@andraxylia  we have 100+ pods. but this is not the reason, I think.

**Comment by jiuchen1986** [comment]
@costinm

**Comment by jiuchen1986** [comment]
@costinm 
Does istio support specifying ports when using https redirection in the gateways? Since I can only find a bool parameter "httpsRedirect" in "Server.TLSOptions", the plain http requests seems to be redirected always to the default https endpoint, which normally listens at 443. If I was right, how can https redirection works when using NodePort to expose the gateways?

**Comment by jiuchen1986** [comment]
@adamzhoul 
Thanks for your advice. Yes we change to a pure ingress-controller solution based on traefik.
But I think if the istio gateway model meets our demands, why not use it cause we have already had mesh in our cluster.

**Comment by jiuchen1986** [comment]
@adamzhoul @costinm

**Comment by jiuchen1986** [comment]
@adamzhoul @costinm 
Ok, so in my case, I can use "hostPort" instead of "hostNetwork" on my gateway pods, and seems it solves my problem.

**Comment by costinm** [comment]
Yes, hostPort should work - the root cause is that Istio currently uses the IP of the pod as a key. 
With hostNetwork, you get the host IP - which is no longer a key ( since ingresstageway and other pods using host networking get the same IP ). 

The reason we use IP as key: readiness/liveness probes and Service. K8S maintains a list of IPs associated with a service, and can remove/add to the list if the pods are healthy. Unfortunately the list (Service.EndpointSet) only has IPs, so we need to match the IPs against pods to find labels.

In 1.1 ( and 1.0.4 ) the sidecars send the full list of labels as Envoy meta - so we can start removing this limitation. At the same time, we are planning to simplify the Gateway selection: instead of using labels and a Gateway in each namespace, we would use the namespace and name of the gateway.

There are many other reasons for the planned changes  - they'll likely fix host networking as well, but short term I would recommend hostPort, which has been tested (it's what we use in minikube tests). It is true that hostPort still goes trough iptables and is less efficient than hostNetwork - but it's also a bit more secure. 

Most common is to use a L4 load balancer in front of Istio ingress.

**Comment by costinm** [comment]
For workarounds:
- removing the labels work by exploiting a validation bug in Istio - it's probably of for 1.0, no plan to fix the bug. ( i.e. a Gateway without selector will match all gateway workloads, regardless of namespace or name)

- if you have a single istio app with 'host' network - it's still going to work - the IP of the host will map to the pod.

- if you have 2 apps with 'host network' - that's where the problem happens, we use a hashmap keyed by IP, and whatever comes last will be in the map.

**Comment by mandarjog** [comment]
I have tested this workaround by removing `selector` from here and we will send that gateway config to all gateway deployments.

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-gateway
  namespace: istio-system
 spec:
  selector:
    istio: ingressgateway   <--- remove this.
  servers:
  - hosts:
    - '*'
```

**Comment by mandarjog** [comment]
@costinm @munrodg what is the situation for this issue in 1.1

**Comment by mdhume** [comment]
@mandarjog @costinm @munrodg Is this fixed in 1.1.x?

**Comment by nishanth-pinnapareddy** [comment]
@mdhume we are facing this issue in 1.1.7 version. Removing the selector on gateway resource mitigated the problem for us !!!

**Comment by howardjohn** [comment]
I think this is fixed in 1.3. We determine service association from node metadata now, not pod IP
